### PR TITLE
bug: crdprovisioner check state

### LINKED ISSUE #179: bug: `crdProvisioner` can trigger redundant requests.
**What happened**:

* It is possible for crdprovisioner to trigger controllers to make duplicate attach / detach / create / delete requests if the previous publishVolume, unpublishVolume, createVolume, deleteVolume requests timed out and was retried by the csi-provisioner or csi-attacher.

**What you expected to happen**:
* crdProvisioner should check CRI states to prevent the abovementioned.

**How to reproduce it**:
* If the volume create / delete / attach / detach operations do not complete within the csi calls' set context timeout, the retried csi request will retrigger volume create / delete / attach / detach by resetting the `Status.State` field to pending state.

**Anything else we need to know?**:

**Environment**:
- CSI Driver version:
- Kubernetes version (use `kubectl version`):
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by sunpa93**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

* It is possible for `crdprovisioner` to trigger controllers to make duplicate attach / detach / create / delete requests if the previous  `publishVolume`, `unpublishVolume`, `createVolume`, `deleteVolume` requests timed out and was retried by the csi-provisioner or csi-attacher.
* `crdProvisioner` should check CRI states to prevent the abovementioned.

**Which issue(s) this PR fixes**:

Fixes #179

**Requirements**:
- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

**Release note**:
```
none
```

### CODE REVIEW

File: pkg/provisioner/crdprovisioner.go [hunk 1]
```diff
@@ -155,29 +155,34 @@ func (c *CrdProvisioner) CreateVolume(
...
-					updateInstance.Spec.CapacityRange = capacityRange
-					updateInstance.Spec.Parameters = parameters
-					updateInstance.Spec.Secrets = secrets
-					updateInstance.Spec.ContentVolumeSource = volumeContentSource
-					updateInstance.Spec.AccessibilityRequirements = accessibilityReq
-				}
+		}
+		// return if volume creation is already in process to prevent duplicate request
+		if azVolumeInstance.Status.State == v1alpha1.VolumeCreating {
+			return nil, status.Errorf(codes.Internal, "creation still in process for volume (%s)", volumeName)
```
Comment (edreed) [inline pkg/provisioner/crdprovisioner.go:161]: Generally, `codes.Aborted` is returned for duplicate operations. See [Container Storage Interface / Spec / Error Scheme](https://github.com/container-storage-interface/spec/blob/master/spec.md#error-scheme)

File: pkg/provisioner/crdprovisioner.go [hunk 2]
```diff
@@ -278,6 +283,12 @@ func (c *CrdProvisioner) DeleteVolume(ctx context.Context, volumeID string, secr
 		return err
 	}
 
+	// if volume deletion already in process, return to prevent duplicate request
+	if azVolume.Status.State == v1alpha1.VolumeDeleting {
+		return status.Errorf(codes.Internal, "deletion still in process for volume (%s)", volumeName)
```
Comment (edreed) [inline pkg/provisioner/crdprovisioner.go:288]: `codes.Aborted`?

Comment (edreed) [reply pkg/provisioner/crdprovisioner.go:288]: Ditto for remaning cases where `codes.Internal` is returned for a duplicate/pending op.

**Comment by edreed** [comment]
Please create a bug issue for this work item, and then change the conventional commit prefix to `bug: ` and reference the item in `Fixes #`. I'd like us to get into the habit of filing issues for all changes going forward to help track and prioritize outstanding work.

**Comment by sunpa93** [comment]
> Please create a bug issue for this work item, and then change the conventional commit prefix to `bug: ` and reference the item in `Fixes #`. I'd like us to get into the habit of filing issues for all changes going forward to help track and prioritize outstanding work.

Done 👍

**Review by edreed — APPROVED** [review]

File: pkg/provisioner/crdprovisioner.go [hunk 3]
```diff
@@ -343,10 +354,18 @@ func (c *CrdProvisioner) PublishVolume(
 	}
 	azVolumeAttachmentInstance, err := azVA.Get(ctx, attachmentName, metav1.GetOptions{})
 	if err == nil {
+		// if CRI is scheduled for deletion, retry until operations are complete
+		if azVolumeAttachmentInstance.DeletionTimestamp != nil {
+			return nil, status.Errorf(codes.Internal, "need to wait until attachment is fully deleted for volume (%s) and node (%s) before attaching", volumeName, nodeID)
```
Comment (edreed) [inline pkg/provisioner/crdprovisioner.go:359]: `codes.Aborted`?

**Pull request merged**